File: NorthwindBlazor/client/Services/NorthwindService.cs
using System;
using System.Web;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Blazor.Components;
using Microsoft.AspNetCore.Blazor.Services;
using NorthwindBlazor.Models.Northwind;

namespace NorthwindBlazor.App
{
    public class NorthwindService 
    {
      public string BasePath { get; set; }

      public ODataService OData { get; set; }

      public NorthwindService()
      {
        OData = new ODataService();
      }
        
      public async Task<ODataServiceResult<Category>> GetCategories(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Category>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Categories", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Category> CreateCategory(Category category)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Categories", category);
      }
    
      public async Task<HttpResponseMessage> DeleteCategory(int? categoryId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Categories({categoryId})");
      }
    
      public async Task<Category> GetCategoryByCategoryId(int? categoryId)
      {
        return await OData.GetBy<Category>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Categories({categoryId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateCategory(int? categoryId, Category category)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Categories({categoryId})", category);
      }
            
      public async Task<ODataServiceResult<Customer>> GetCustomers(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Customer>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Customers", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Customer> CreateCustomer(Customer customer)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Customers", customer);
      }
    
      public async Task<HttpResponseMessage> DeleteCustomer(string customerId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Customers('{HttpUtility.UrlEncode(customerId)}')");
      }
    
      public async Task<Customer> GetCustomerByCustomerId(string customerId)
      {
        return await OData.GetBy<Customer>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Customers('{HttpUtility.UrlEncode(customerId)}')" });
      }
    
      public async Task<HttpResponseMessage> UpdateCustomer(string customerId, Customer customer)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Customers('{HttpUtility.UrlEncode(customerId)}')", customer);
      }
            
      public async Task<ODataServiceResult<CustomerCustomerDemo>> GetCustomerCustomerDemos(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<CustomerCustomerDemo>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/CustomerCustomerDemos", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<CustomerCustomerDemo> CreateCustomerCustomerDemo(CustomerCustomerDemo customerCustomerDemo)
      {
        return await OData.Add($"{BasePath}odata/Northwind/CustomerCustomerDemos", customerCustomerDemo);
      }
            
      public async Task<ODataServiceResult<CustomerDemographic>> GetCustomerDemographics(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<CustomerDemographic>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/CustomerDemographics", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<CustomerDemographic> CreateCustomerDemographic(CustomerDemographic customerDemographic)
      {
        return await OData.Add($"{BasePath}odata/Northwind/CustomerDemographics", customerDemographic);
      }
    
      public async Task<HttpResponseMessage> DeleteCustomerDemographic(string customerTypeId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/CustomerDemographics('{HttpUtility.UrlEncode(customerTypeId)}')");
      }
    
      public async Task<CustomerDemographic> GetCustomerDemographicByCustomerTypeId(string customerTypeId)
      {
        return await OData.GetBy<CustomerDemographic>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/CustomerDemographics('{HttpUtility.UrlEncode(customerTypeId)}')" });
      }
    
      public async Task<HttpResponseMessage> UpdateCustomerDemographic(string customerTypeId, CustomerDemographic customerDemographic)
      {
        return await OData.Update($"{BasePath}odata/Northwind/CustomerDemographics('{HttpUtility.UrlEncode(customerTypeId)}')", customerDemographic);
      }
            
      public async Task<ODataServiceResult<Employee>> GetEmployees(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Employee>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Employees", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Employee> CreateEmployee(Employee employee)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Employees", employee);
      }
    
      public async Task<HttpResponseMessage> DeleteEmployee(int? employeeId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Employees({employeeId})");
      }
    
      public async Task<Employee> GetEmployeeByEmployeeId(int? employeeId)
      {
        return await OData.GetBy<Employee>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Employees({employeeId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateEmployee(int? employeeId, Employee employee)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Employees({employeeId})", employee);
      }
            
      public async Task<ODataServiceResult<EmployeeTerritory>> GetEmployeeTerritories(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<EmployeeTerritory>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/EmployeeTerritories", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<EmployeeTerritory> CreateEmployeeTerritory(EmployeeTerritory employeeTerritory)
      {
        return await OData.Add($"{BasePath}odata/Northwind/EmployeeTerritories", employeeTerritory);
      }
            
      public async Task<ODataServiceResult<Order>> GetOrders(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Order>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Orders", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Order> CreateOrder(Order order)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Orders", order);
      }
    
      public async Task<HttpResponseMessage> DeleteOrder(int? orderId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Orders({orderId})");
      }
    
      public async Task<Order> GetOrderByOrderId(int? orderId)
      {
        return await OData.GetBy<Order>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Orders({orderId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateOrder(int? orderId, Order order)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Orders({orderId})", order);
      }
            
      public async Task<ODataServiceResult<OrderDetail>> GetOrderDetails(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<OrderDetail>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/OrderDetails", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<OrderDetail> CreateOrderDetail(OrderDetail orderDetail)
      {
        return await OData.Add($"{BasePath}odata/Northwind/OrderDetails", orderDetail);
      }
            
      public async Task<ODataServiceResult<Product>> GetProducts(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Product>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Products", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Product> CreateProduct(Product product)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Products", product);
      }
    
      public async Task<HttpResponseMessage> DeleteProduct(int? productId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Products({productId})");
      }
    
      public async Task<Product> GetProductByProductId(int? productId)
      {
        return await OData.GetBy<Product>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Products({productId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateProduct(int? productId, Product product)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Products({productId})", product);
      }
            
      public async Task<ODataServiceResult<Region>> GetRegions(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Region>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Regions", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Region> CreateRegion(Region region)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Regions", region);
      }
    
      public async Task<HttpResponseMessage> DeleteRegion(int? regionId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Regions({regionId})");
      }
    
      public async Task<Region> GetRegionByRegionId(int? regionId)
      {
        return await OData.GetBy<Region>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Regions({regionId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateRegion(int? regionId, Region region)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Regions({regionId})", region);
      }
            
      public async Task<ODataServiceResult<Shipper>> GetShippers(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Shipper>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Shippers", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Shipper> CreateShipper(Shipper shipper)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Shippers", shipper);
      }
    
      public async Task<HttpResponseMessage> DeleteShipper(int? shipperId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Shippers({shipperId})");
      }
    
      public async Task<Shipper> GetShipperByShipperId(int? shipperId)
      {
        return await OData.GetBy<Shipper>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Shippers({shipperId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateShipper(int? shipperId, Shipper shipper)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Shippers({shipperId})", shipper);
      }
            
      public async Task<ODataServiceResult<Supplier>> GetSuppliers(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Supplier>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Suppliers", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Supplier> CreateSupplier(Supplier supplier)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Suppliers", supplier);
      }
    
      public async Task<HttpResponseMessage> DeleteSupplier(int? supplierId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Suppliers({supplierId})");
      }
    
      public async Task<Supplier> GetSupplierBySupplierId(int? supplierId)
      {
        return await OData.GetBy<Supplier>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Suppliers({supplierId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateSupplier(int? supplierId, Supplier supplier)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Suppliers({supplierId})", supplier);
      }
            
      public async Task<ODataServiceResult<Territory>> GetTerritories(string filter, int? top, int? skip, string orderby, bool? count, string expand, string format, string select)
      {
        return await OData.Get<Territory>(new ODataServiceArgs() { Url = $"{BasePath}odata/Northwind/Territories", Filter = filter, Top = top, Skip = skip, Orderby = orderby, Count = count, Expand = expand, Format = format, Select = select });
      }
    
      public async Task<Territory> CreateTerritory(Territory territory)
      {
        return await OData.Add($"{BasePath}odata/Northwind/Territories", territory);
      }
    
      public async Task<HttpResponseMessage> DeleteTerritory(string territoryId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/Territories('{HttpUtility.UrlEncode(territoryId)}')");
      }
    
      public async Task<Territory> GetTerritoryByTerritoryId(string territoryId)
      {
        return await OData.GetBy<Territory>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/Territories('{HttpUtility.UrlEncode(territoryId)}')" });
      }
    
      public async Task<HttpResponseMessage> UpdateTerritory(string territoryId, Territory territory)
      {
        return await OData.Update($"{BasePath}odata/Northwind/Territories('{HttpUtility.UrlEncode(territoryId)}')", territory);
      }
            
      public async Task<HttpResponseMessage> DeleteCustomerCustomerDemo(string customerId, string customerTypeId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/CustomerCustomerDemos(CustomerID='{HttpUtility.UrlEncode(customerId)}',CustomerTypeID='{HttpUtility.UrlEncode(customerTypeId)}')");
      }
    
      public async Task<CustomerCustomerDemo> GetCustomerCustomerDemoByCustomerIdAndCustomerTypeId(string customerId, string customerTypeId)
      {
        return await OData.GetBy<CustomerCustomerDemo>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/CustomerCustomerDemos(CustomerID='{HttpUtility.UrlEncode(customerId)}',CustomerTypeID='{HttpUtility.UrlEncode(customerTypeId)}')" });
      }
    
      public async Task<HttpResponseMessage> UpdateCustomerCustomerDemo(string customerId, string customerTypeId, CustomerCustomerDemo customerCustomerDemo)
      {
        return await OData.Update($"{BasePath}odata/Northwind/CustomerCustomerDemos(CustomerID='{HttpUtility.UrlEncode(customerId)}',CustomerTypeID='{HttpUtility.UrlEncode(customerTypeId)}')", customerCustomerDemo);
      }
            
      public async Task<HttpResponseMessage> DeleteEmployeeTerritory(int? employeeId, string territoryId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/EmployeeTerritories(EmployeeID={employeeId},TerritoryID='{HttpUtility.UrlEncode(territoryId)}')");
      }
    
      public async Task<EmployeeTerritory> GetEmployeeTerritoryByEmployeeIdAndTerritoryId(int? employeeId, string territoryId)
      {
        return await OData.GetBy<EmployeeTerritory>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/EmployeeTerritories(EmployeeID={employeeId},TerritoryID='{HttpUtility.UrlEncode(territoryId)}')" });
      }
    
      public async Task<HttpResponseMessage> UpdateEmployeeTerritory(int? employeeId, string territoryId, EmployeeTerritory employeeTerritory)
      {
        return await OData.Update($"{BasePath}odata/Northwind/EmployeeTerritories(EmployeeID={employeeId},TerritoryID='{HttpUtility.UrlEncode(territoryId)}')", employeeTerritory);
      }
            
      public async Task<HttpResponseMessage> DeleteOrderDetail(int? orderId, int? productId)
      {
        return await OData.Delete($"{BasePath}odata/Northwind/OrderDetails(OrderID={orderId},ProductID={productId})");
      }
    
      public async Task<OrderDetail> GetOrderDetailByOrderIdAndProductId(int? orderId, int? productId)
      {
        return await OData.GetBy<OrderDetail>(new ODataServiceItemArgs() { Url = $"{BasePath}odata/Northwind/OrderDetails(OrderID={orderId},ProductID={productId})" });
      }
    
      public async Task<HttpResponseMessage> UpdateOrderDetail(int? orderId, int? productId, OrderDetail orderDetail)
      {
        return await OData.Update($"{BasePath}odata/Northwind/OrderDetails(OrderID={orderId},ProductID={productId})", orderDetail);
      }
        
  }
}

File: NorthwindDashboard/client/src/app/northwind.service.ts
import { Injectable } from '@angular/core';
import { Location } from '@angular/common';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable, BehaviorSubject, throwError } from 'rxjs';

import { ConfigService } from './config.service';
import { ODataClient } from './odata-client';
import * as models from './northwind.model';

@Injectable()
export class NorthwindService {
  odata: ODataClient;
  basePath: string;

  constructor(private http: HttpClient, private config: ConfigService) {
    this.basePath = config.get('northwind');
    this.odata = new ODataClient(this.http, this.basePath, { legacy: false, withCredentials: true });
  }

  getCategories(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Categories`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createCategory(expand: string | null, category: models.Category | null) : Observable<any> {
    return this.odata.post(`/Categories`, category, { expand }, []);
  }

  deleteCategory(categoryId: number | null) : Observable<any> {
    return this.odata.delete(`/Categories(${categoryId})`, item => !(item.CategoryID == categoryId));
  }

  getCategoryByCategoryId(expand: string | null, categoryId: number | null) : Observable<any> {
    return this.odata.getById(`/Categories(${categoryId})`, { expand });
  }

  updateCategory(expand: string | null, categoryId: number | null, category: models.Category | null) : Observable<any> {
    return this.odata.patch(`/Categories(${categoryId})`, category, item => item.CategoryID == categoryId, { expand }, []);
  }

  getCustomers(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Customers`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createCustomer(expand: string | null, customer: models.Customer | null) : Observable<any> {
    return this.odata.post(`/Customers`, customer, { expand }, []);
  }

  deleteCustomer(customerId: string | null) : Observable<any> {
    return this.odata.delete(`/Customers('${encodeURIComponent(customerId)}')`, item => !(item.CustomerID == customerId));
  }

  getCustomerByCustomerId(expand: string | null, customerId: string | null) : Observable<any> {
    return this.odata.getById(`/Customers('${encodeURIComponent(customerId)}')`, { expand });
  }

  updateCustomer(expand: string | null, customerId: string | null, customer: models.Customer | null) : Observable<any> {
    return this.odata.patch(`/Customers('${encodeURIComponent(customerId)}')`, customer, item => item.CustomerID == customerId, { expand }, []);
  }

  getCustomerCustomerDemos(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/CustomerCustomerDemos`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createCustomerCustomerDemo(expand: string | null, customerCustomerDemo: models.CustomerCustomerDemo | null) : Observable<any> {
    return this.odata.post(`/CustomerCustomerDemos`, customerCustomerDemo, { expand }, ['Customer', 'CustomerDemographic']);
  }

  getCustomerDemographics(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/CustomerDemographics`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createCustomerDemographic(expand: string | null, customerDemographic: models.CustomerDemographic | null) : Observable<any> {
    return this.odata.post(`/CustomerDemographics`, customerDemographic, { expand }, []);
  }

  deleteCustomerDemographic(customerTypeId: string | null) : Observable<any> {
    return this.odata.delete(`/CustomerDemographics('${encodeURIComponent(customerTypeId)}')`, item => !(item.CustomerTypeID == customerTypeId));
  }

  getCustomerDemographicByCustomerTypeId(expand: string | null, customerTypeId: string | null) : Observable<any> {
    return this.odata.getById(`/CustomerDemographics('${encodeURIComponent(customerTypeId)}')`, { expand });
  }

  updateCustomerDemographic(expand: string | null, customerTypeId: string | null, customerDemographic: models.CustomerDemographic | null) : Observable<any> {
    return this.odata.patch(`/CustomerDemographics('${encodeURIComponent(customerTypeId)}')`, customerDemographic, item => item.CustomerTypeID == customerTypeId, { expand }, []);
  }

  getEmployees(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Employees`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createEmployee(expand: string | null, employee: models.Employee | null) : Observable<any> {
    return this.odata.post(`/Employees`, employee, { expand }, ['Employee1']);
  }

  deleteEmployee(employeeId: number | null) : Observable<any> {
    return this.odata.delete(`/Employees(${employeeId})`, item => !(item.EmployeeID == employeeId));
  }

  getEmployeeByEmployeeId(expand: string | null, employeeId: number | null) : Observable<any> {
    return this.odata.getById(`/Employees(${employeeId})`, { expand });
  }

  updateEmployee(expand: string | null, employeeId: number | null, employee: models.Employee | null) : Observable<any> {
    return this.odata.patch(`/Employees(${employeeId})`, employee, item => item.EmployeeID == employeeId, { expand }, ['Employee1']);
  }

  getEmployeeTerritories(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/EmployeeTerritories`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createEmployeeTerritory(expand: string | null, employeeTerritory: models.EmployeeTerritory | null) : Observable<any> {
    return this.odata.post(`/EmployeeTerritories`, employeeTerritory, { expand }, ['Employee', 'Territory']);
  }

  getOrders(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Orders`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createOrder(expand: string | null, order: models.Order | null) : Observable<any> {
    return this.odata.post(`/Orders`, order, { expand }, ['Customer', 'Employee', 'Shipper']);
  }

  deleteOrder(orderId: number | null) : Observable<any> {
    return this.odata.delete(`/Orders(${orderId})`, item => !(item.OrderID == orderId));
  }

  getOrderByOrderId(expand: string | null, orderId: number | null) : Observable<any> {
    return this.odata.getById(`/Orders(${orderId})`, { expand });
  }

  updateOrder(expand: string | null, orderId: number | null, order: models.Order | null) : Observable<any> {
    return this.odata.patch(`/Orders(${orderId})`, order, item => item.OrderID == orderId, { expand }, ['Customer','Employee','Shipper']);
  }

  getOrderDetails(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/OrderDetails`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createOrderDetail(expand: string | null, orderDetail: models.OrderDetail | null) : Observable<any> {
    return this.odata.post(`/OrderDetails`, orderDetail, { expand }, ['Order', 'Product']);
  }

  getProducts(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Products`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createProduct(expand: string | null, product: models.Product | null) : Observable<any> {
    return this.odata.post(`/Products`, product, { expand }, ['Supplier', 'Category']);
  }

  deleteProduct(productId: number | null) : Observable<any> {
    return this.odata.delete(`/Products(${productId})`, item => !(item.ProductID == productId));
  }

  getProductByProductId(expand: string | null, productId: number | null) : Observable<any> {
    return this.odata.getById(`/Products(${productId})`, { expand });
  }

  updateProduct(expand: string | null, productId: number | null, product: models.Product | null) : Observable<any> {
    return this.odata.patch(`/Products(${productId})`, product, item => item.ProductID == productId, { expand }, ['Supplier','Category']);
  }

  getRegions(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Regions`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createRegion(expand: string | null, region: models.Region | null) : Observable<any> {
    return this.odata.post(`/Regions`, region, { expand }, []);
  }

  deleteRegion(regionId: number | null) : Observable<any> {
    return this.odata.delete(`/Regions(${regionId})`, item => !(item.RegionID == regionId));
  }

  getRegionByRegionId(expand: string | null, regionId: number | null) : Observable<any> {
    return this.odata.getById(`/Regions(${regionId})`, { expand });
  }

  updateRegion(expand: string | null, regionId: number | null, region: models.Region | null) : Observable<any> {
    return this.odata.patch(`/Regions(${regionId})`, region, item => item.RegionID == regionId, { expand }, []);
  }

  getShippers(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Shippers`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createShipper(expand: string | null, shipper: models.Shipper | null) : Observable<any> {
    return this.odata.post(`/Shippers`, shipper, { expand }, []);
  }

  deleteShipper(shipperId: number | null) : Observable<any> {
    return this.odata.delete(`/Shippers(${shipperId})`, item => !(item.ShipperID == shipperId));
  }

  getShipperByShipperId(expand: string | null, shipperId: number | null) : Observable<any> {
    return this.odata.getById(`/Shippers(${shipperId})`, { expand });
  }

  updateShipper(expand: string | null, shipperId: number | null, shipper: models.Shipper | null) : Observable<any> {
    return this.odata.patch(`/Shippers(${shipperId})`, shipper, item => item.ShipperID == shipperId, { expand }, []);
  }

  getSuppliers(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Suppliers`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createSupplier(expand: string | null, supplier: models.Supplier | null) : Observable<any> {
    return this.odata.post(`/Suppliers`, supplier, { expand }, []);
  }

  deleteSupplier(supplierId: number | null) : Observable<any> {
    return this.odata.delete(`/Suppliers(${supplierId})`, item => !(item.SupplierID == supplierId));
  }

  getSupplierBySupplierId(expand: string | null, supplierId: number | null) : Observable<any> {
    return this.odata.getById(`/Suppliers(${supplierId})`, { expand });
  }

  updateSupplier(expand: string | null, supplierId: number | null, supplier: models.Supplier | null) : Observable<any> {
    return this.odata.patch(`/Suppliers(${supplierId})`, supplier, item => item.SupplierID == supplierId, { expand }, []);
  }

  getTerritories(filter: string | null, top: number | null, skip: number | null, orderby: string | null, count: boolean | null, expand: string | null, format: string | null, select: string | null) : Observable<any> {
    return this.odata.get(`/Territories`, { filter, top, skip, orderby, count, expand, format, select });
  }

  createTerritory(expand: string | null, territory: models.Territory | null) : Observable<any> {
    return this.odata.post(`/Territories`, territory, { expand }, ['Region']);
  }

  deleteTerritory(territoryId: string | null) : Observable<any> {
    return this.odata.delete(`/Territories('${encodeURIComponent(territoryId)}')`, item => !(item.TerritoryID == territoryId));
  }

  getTerritoryByTerritoryId(expand: string | null, territoryId: string | null) : Observable<any> {
    return this.odata.getById(`/Territories('${encodeURIComponent(territoryId)}')`, { expand });
  }

  updateTerritory(expand: string | null, territoryId: string | null, territory: models.Territory | null) : Observable<any> {
    return this.odata.patch(`/Territories('${encodeURIComponent(territoryId)}')`, territory, item => item.TerritoryID == territoryId, { expand }, ['Region']);
  }

  deleteCustomerCustomerDemo(customerId: string | null, customerTypeId: string | null) : Observable<any> {
    return this.odata.delete(`/CustomerCustomerDemos(CustomerID='${encodeURIComponent(customerId)}',CustomerTypeID='${encodeURIComponent(customerTypeId)}')`, item => !(item.CustomerID == customerId && item.CustomerTypeID == customerTypeId));
  }

  getCustomerCustomerDemoByCustomerIdAndCustomerTypeId(customerId: string | null, customerTypeId: string | null, expand: string | null) : Observable<any> {
    return this.odata.getById(`/CustomerCustomerDemos(CustomerID='${encodeURIComponent(customerId)}',CustomerTypeID='${encodeURIComponent(customerTypeId)}')`, { expand });
  }

  updateCustomerCustomerDemo(customerId: string | null, customerTypeId: string | null, expand: string | null, customerCustomerDemo: models.CustomerCustomerDemo | null) : Observable<any> {
    return this.odata.patch(`/CustomerCustomerDemos(CustomerID='${encodeURIComponent(customerId)}',CustomerTypeID='${encodeURIComponent(customerTypeId)}')`, customerCustomerDemo, item => item.CustomerID == customerId && item.CustomerTypeID == customerTypeId, { expand }, ['Customer','CustomerDemographic']);
  }

  deleteEmployeeTerritory(employeeId: number | null, territoryId: string | null) : Observable<any> {
    return this.odata.delete(`/EmployeeTerritories(EmployeeID=${employeeId},TerritoryID='${encodeURIComponent(territoryId)}')`, item => !(item.EmployeeID == employeeId && item.TerritoryID == territoryId));
  }

  getEmployeeTerritoryByEmployeeIdAndTerritoryId(employeeId: number | null, territoryId: string | null, expand: string | null) : Observable<any> {
    return this.odata.getById(`/EmployeeTerritories(EmployeeID=${employeeId},TerritoryID='${encodeURIComponent(territoryId)}')`, { expand });
  }

  updateEmployeeTerritory(employeeId: number | null, territoryId: string | null, expand: string | null, employeeTerritory: models.EmployeeTerritory | null) : Observable<any> {
    return this.odata.patch(`/EmployeeTerritories(EmployeeID=${employeeId},TerritoryID='${encodeURIComponent(territoryId)}')`, employeeTerritory, item => item.EmployeeID == employeeId && item.TerritoryID == territoryId, { expand }, ['Employee','Territory']);
  }

  deleteOrderDetail(orderId: number | null, productId: number | null) : Observable<any> {
    return this.odata.delete(`/OrderDetails(OrderID=${orderId},ProductID=${productId})`, item => !(item.OrderID == orderId && item.ProductID == productId));
  }

  getOrderDetailByOrderIdAndProductId(orderId: number | null, productId: number | null, expand: string | null) : Observable<any> {
    return this.odata.getById(`/OrderDetails(OrderID=${orderId},ProductID=${productId})`, { expand });
  }

  updateOrderDetail(orderId: number | null, productId: number | null, expand: string | null, orderDetail: models.OrderDetail | null) : Observable<any> {
    return this.odata.patch(`/OrderDetails(OrderID=${orderId},ProductID=${productId})`, orderDetail, item => item.OrderID == orderId && item.ProductID == productId, { expand }, ['Order','Product']);
  }
}
